File: LiftLog.Ui/Services/SessionService.cs
using System.Collections.Immutable;
using Fluxor;
using LiftLog.Lib;
using LiftLog.Lib.Models;
using LiftLog.Ui.Store.CurrentSession;
using LiftLog.Ui.Store.Settings;

namespace LiftLog.Ui.Services;

public class SessionService(
    IState<SettingsState> settingsState,
    IState<CurrentSessionState> currentSessionState,
    ProgressRepository progressRepository
)
{
    /// <summary>
    /// Returns all future sessions in order, including the current session if there is one.
    /// This enumerable is infinite, so ensure to limit output when consuming.
    /// </summary>
    public async IAsyncEnumerable<Session> GetUpcomingSessionsAsync(
        ImmutableListValue<SessionBlueprint> sessionBlueprints
    )
    {
        if (!sessionBlueprints.Any())
        {
            yield break;
        }

        var latestRecordedExercises = await progressRepository.GetLatestRecordedExercisesAsync();
        var currentSession = currentSessionState.Value.WorkoutSession;
        if (currentSession?.IsStarted == true)
            yield return currentSession;

        var latestSession = currentSession switch
        {
            { IsStarted: true } => currentSession,
            _ => await progressRepository
                .GetOrderedSessions()
                .FirstOrDefaultAsync(x => !x.IsFreeform),
        };
        if (latestSession == null)
        {
            latestSession = CreateNewSession(sessionBlueprints[0], latestRecordedExercises);
            yield return latestSession;
        }

        // We need the blueprint that comes after this session
        while (true)
        {
            latestSession = GetNextSession(
                latestSession,
                sessionBlueprints,
                latestRecordedExercises
            );
            yield return latestSession;
        }
    }

    public IAsyncEnumerable<Session> GetLatestSessionsAsync()
    {
        return progressRepository.GetOrderedSessions();
    }

    public async Task<Session> HydrateSessionFromBlueprint(SessionBlueprint sessionBlueprint)
    {
        var latestRecordedExercises = await progressRepository.GetLatestRecordedExercisesAsync();
        return CreateNewSession(sessionBlueprint, latestRecordedExercises);
    }

    private Session GetNextSession(
        Session previousSession,
        ImmutableList<SessionBlueprint> sessionBlueprints,
        IReadOnlyDictionary<KeyedExerciseBlueprint, RecordedExercise> latestRecordedExercises
    )
    {
        var lastBlueprint = previousSession.Blueprint;
        var lastBlueprintIndex = sessionBlueprints.FindIndex(x => x.Name == lastBlueprint.Name);
        return CreateNewSession(
            sessionBlueprints[(lastBlueprintIndex + 1) % sessionBlueprints.Count],
            latestRecordedExercises
        ) with
        {
            Bodyweight = previousSession.Bodyweight,
        };
    }

    private Session CreateNewSession(
        SessionBlueprint sessionBlueprint,
        IReadOnlyDictionary<KeyedExerciseBlueprint, RecordedExercise> latestRecordedExercises
    )
    {
        var splitWeightByDefault = settingsState.Value.SplitWeightByDefault;
        RecordedExercise GetNextExercise(ExerciseBlueprint e)
        {
            var lastExercise = latestRecordedExercises.GetValueOrDefault(e);
            var potentialSets = lastExercise switch
            {
                null or { PerSetWeight: false } => Enumerable.Repeat(
                    new PotentialSet(
                        null,
                        lastExercise?.PotentialSets.FirstOrDefault()?.Weight ?? 0
                    ),
                    e.Sets
                ),
                { IsSuccessForProgressiveOverload: true } => lastExercise.PotentialSets.Select(
                    x => new PotentialSet(null, x.Weight + e.WeightIncreaseOnSuccess)
                ),
                _ => lastExercise.PotentialSets.Select(x => new PotentialSet(null, x.Weight)),
            };
            return new RecordedExercise(
                e,
                potentialSets.ToImmutableList(),
                null,
                splitWeightByDefault || (lastExercise?.PerSetWeight ?? false)
            );
        }

        return new Session(
            Guid.NewGuid(),
            sessionBlueprint,
            sessionBlueprint.Exercises.Select(GetNextExercise).ToImmutableList(),
            DateOnly.FromDateTime(DateTime.Now),
            null
        );
    }
}

File: liftlog-react/services/session-service.ts
import {
  ExerciseBlueprint,
  KeyedExerciseBlueprint,
  SessionBlueprint,
} from '@/models/blueprint-models';
import {
  PotentialSet,
  RecordedExercise,
  Session,
} from '@/models/session-models';
import { ProgressRepository } from '@/services/progress-repository';
import type { RootState } from '@/store';
import { uuid } from '@/utils/uuid';
import { LocalDate } from '@js-joda/core';
import BigNumber from 'bignumber.js';
import Enumerable from 'linq';
import { match, P } from 'ts-pattern';

export class SessionService {
  constructor(
    private progressRepository: ProgressRepository,
    private getState: () => RootState,
  ) {}

  async *getUpcomingSessions(
    sessionBlueprints: SessionBlueprint[],
  ): AsyncIterableIterator<Session> {
    const currentState = this.getState();
    const currentSession = Session.fromPOJO(
      currentState.currentSession.workoutSession,
    );
    if (currentSession?.isStarted) {
      yield currentSession;
    }

    if (!sessionBlueprints.length) {
      return;
    }
    // TODO: does not match original impl

    const latestRecordedExercises =
      await this.progressRepository.getLatestRecordedExercises();
    let latestSession = await match(currentSession)
      .with({ isStarted: true }, (x) => x)
      .otherwise(async () =>
        (await this.progressRepository.getOrderedSessions()).firstOrDefault(
          (x) => !x.isFreeform,
        ),
      );
    if (!latestSession) {
      latestSession = this.createNewSession(
        sessionBlueprints[0],
        latestRecordedExercises,
      );
      yield latestSession;
    }

    while (true) {
      latestSession = this.getNextSession(
        latestSession,
        sessionBlueprints,
        latestRecordedExercises,
      );
      yield latestSession;
    }
  }

  private getNextSession(
    previousSession: Session,
    sessionBlueprints: SessionBlueprint[],
    latestRecordedExercises: Enumerable.IDictionary<
      KeyedExerciseBlueprint,
      RecordedExercise
    >,
  ): Session {
    const lastBlueprint = previousSession.blueprint;
    const lastBlueprintIndex = sessionBlueprints.findIndex(
      (x) => x.name === lastBlueprint.name,
    );
    const nextBlueprint =
      sessionBlueprints[(lastBlueprintIndex + 1) % sessionBlueprints.length];

    return this.createNewSession(nextBlueprint, latestRecordedExercises).with({
      bodyweight: previousSession.bodyweight,
    });
  }

  private createNewSession(
    sessionBlueprint: SessionBlueprint,
    latestRecordedExercises: Enumerable.IDictionary<
      KeyedExerciseBlueprint,
      RecordedExercise
    >,
  ): Session {
    const {
      settings: { splitWeightByDefault },
    } = this.getState();
    function getNextExercise(e: ExerciseBlueprint): RecordedExercise {
      const lastExercise = latestRecordedExercises.get(e);
      const potentialSets: PotentialSet[] = match(lastExercise)
        .returnType<PotentialSet[]>()
        .with(
          P.union(undefined, {
            perSetWeight: false,
            isSuccessForProgressiveOverload: false,
          }),
          () =>
            Array.from({ length: e.sets }, () =>
              PotentialSet.fromPOJO({
                weight: lastExercise?.potentialSets[0]?.weight ?? BigNumber(0),
                set: undefined,
              }),
            ),
        )
        .with({ isSuccessForProgressiveOverload: true }, (x) =>
          x.potentialSets.map((x) =>
            PotentialSet.fromPOJO({
              weight: x.weight.plus(e.weightIncreaseOnSuccess),
              set: undefined,
            }),
          ),
        )
        .otherwise((x) =>
          x.potentialSets.map((x) =>
            PotentialSet.fromPOJO({
              weight: x.weight,
              set: undefined,
            }),
          ),
        );

      return new RecordedExercise(
        e,
        potentialSets,
        undefined,
        splitWeightByDefault || (lastExercise?.perSetWeight ?? false),
      );
    }
    return new Session(
      uuid(),
      sessionBlueprint,
      sessionBlueprint.exercises.map(getNextExercise),
      LocalDate.now(),
      undefined,
    );
  }
}
